File: backend/src/main/java/com/staffdesk/ems/attendance/entity/Attendance.java
package com.staffdesk.ems.attendance.entity;

import com.staffdesk.ems.employee.entity.Employee;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.Instant;
import java.time.LocalDate;

@Entity
@Table(name = "attendance")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Attendance {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "employee_id", nullable = false)
    private Employee employee;

    @Column(name = "attendance_date", nullable = false)
    private LocalDate attendanceDate;

    @Column(name = "clock_in")
    private Instant clockIn;

    @Column(name = "clock_out")
    private Instant clockOut;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private Status status = Status.PRESENT;

    // Note: the attendance table has no updated_at column (unlike employees/departments) —
    // it's an append/upsert-per-day log, not an editable master record.
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @PrePersist
    protected void onCreate() {
        if (createdAt == null) {
            createdAt = Instant.now();
        }
        if (attendanceDate == null) {
            attendanceDate = LocalDate.now();
        }
    }

    public enum Status {
        PRESENT, ABSENT, HALF_DAY, LATE
    }
}

File: frontend/types/attendance.ts
// Mirrors AttendanceResponse.java / AttendanceUpsertRequest.java
import { Page } from "@/types/employee";

export type AttendanceStatus = "PRESENT" | "ABSENT" | "HALF_DAY" | "LATE";

export interface AttendanceRecord {
  id: number;
  employeeId: number;
  employeeCode: string;
  employeeName: string;
  departmentName: string | null;
  attendanceDate: string; // ISO date, e.g. "2026-07-25"
  clockIn: string | null; // ISO instant, e.g. "2026-07-25T09:03:12Z"
  clockOut: string | null;
  status: AttendanceStatus;
  createdAt: string;
}

// For PUT /attendance/employees/{employeeId}/{date} (HR manual create/override)
export interface AttendanceUpsertRequest {
  clockIn: string | null; // ISO instant or null
  clockOut: string | null;
  status: AttendanceStatus;
}

export type AttendancePage = Page<AttendanceRecord>;

export const ATTENDANCE_STATUS_LABEL: Record<AttendanceStatus, string> = {
  PRESENT: "Present",
  ABSENT: "Absent",
  HALF_DAY: "Half day",
  LATE: "Late",
};